File: Python/addtime.py
#!/usr/bin/env python3


class Time:
    """HH:MM simplistic time"""

    def __init__(self, hrs=0, mins=0):
        self.hrs = hrs
        self.mins = mins

    def __repr__(self):
        return f"{self.hrs:02d}:{self.mins:02d}"

    def __add__(self, t):
        """Return Time + t, where t is another Time or Int mins"""
        tot = t + abs(self)
        return Time.from_int(tot)

    def __abs__(self):
        """Returns the total minutes represented"""
        return (self.hrs * 60) + self.mins

    @classmethod
    def from_int(cls, i):
        return cls(*divmod(i, 60))

    @classmethod
    def from_string(cls, string):
        if ':' in string:
            parts = [int(x) for x in string.split(':')]
            return cls(parts[0], parts[1])
        else:
            return cls.from_int(int(string))


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("Start", type=Time.from_string,
                        help="HH:MM or mins, representing beginning time")
    parser.add_argument("Elapse", type=Time.from_string,
                        help="HH:MM or mins, representing time to elapse by")
    parser.add_argument('-q', '--quiet', action='store_true',
                        help="Print elapsed time only")
    args = parser.parse_args()

    elapse = abs(args.Elapse)
    if args.quiet:
        print(f"{args.Start + elapse}")
    else:
        print(f"Start:\t{args.Start}\t{elapse:+}\nEnd:\t{args.Start + elapse}")

File: Ruby/addtime.rb
#!/usr/bin/ruby

class Time
    attr_reader :hrs, :mins
    attr_writer :hrs, :mins
    def initialize(*args)
        if args.length == 1
            if args[0].is_a? String
                tmp = args[0].to_t
                @hrs = tmp.hrs
                @mins = tmp.mins
            else
                absmins = args[0]
                @hrs = (absmins / 60).floor
                @mins = absmins % 60
            end
        else
            @hrs =  args[0] or 0
            @mins = args[1] or 0
        end
    end

    def +(other)
        if other.is_a?(Integer)
            o = other
        elsif other.is_a?(Time)
            o = other.abs
        end
        # not defined <<= 0
        o ||= 0
        return Time.new(self.abs + o)
    end

    def to_s
        "%02d:%02d" % [self.hrs , self.mins]
    end

    def abs
        (60 * self.hrs) + self.mins
    end
end

class String
    def is_i?
        true if Integer(self) rescue false
    end

    def to_t
        if self.is_i?
            return Time.new(self.to_i)
        elsif m = (/(\d{1,2}):(\d{1,2})/.match self)
            return Time.new(m[1].to_i, m[2].to_i)
        end
    end
end

quiet, rest = ARGV.to_enum.partition {|x| x =~ /-/}

if rest.length < 2
    puts "Usage: #{$0} [-q] START ELAPSE"
    puts "\t-q: print elapsed time only"
    exit 1
end

start = rest[0].to_t or Time.new()
elapse = rest[1].to_t or Time.new()
if quiet.length > 0
    puts start + elapse
else
    puts "#{start} + #{elapse.abs} -> #{start + elapse}"
end